File: UnityMicroFund/unitymicrofund_web/src/app/core/services/sms.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap, catchError, throwError } from 'rxjs';

export interface SmsResponse {
  uid: string;
  status: string;
  cost?: number;
}

export interface SmsConfig {
  apiUrl: string;
  senderId: string;
  apiKey: string;
}

@Injectable({
  providedIn: 'root',
})
export class SmsService {
  private readonly defaultConfig: SmsConfig = {
    apiUrl: 'https://app.seasms.com/api/v3',
    senderId: 'UnityMF',
    apiKey: '',
  };

  private config: SmsConfig = this.defaultConfig;

  constructor(private http: HttpClient) {
    this.loadConfig();
  }

  private loadConfig(): void {
    const savedConfig = localStorage.getItem('sms_config');
    if (savedConfig) {
      try {
        this.config = { ...this.defaultConfig, ...JSON.parse(savedConfig) };
      } catch {
        this.config = this.defaultConfig;
      }
    }
  }

  configure(config: Partial<SmsConfig>): void {
    this.config = { ...this.defaultConfig, ...config };
    localStorage.setItem('sms_config', JSON.stringify(this.config));
  }

  sendSms(to: string, message: string): Observable<SmsResponse> {
    const formattedNumber = this.formatBangladeshNumber(to);
    
    return this.http.post<SmsResponse>(`${this.config.apiUrl}/sms/send`, {
      recipient: formattedNumber,
      sender_id: this.config.senderId,
      type: 'plain',
      message: message,
    }, {
      headers: {
        'Authorization': `Bearer ${this.config.apiKey}`,
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    }).pipe(
      tap((response) => {
        console.log('SMS sent successfully:', response);
      }),
      catchError((error) => {
        console.error('SMS send error:', error);
        return throwError(() => error);
      })
    );
  }

  sendOtp(phoneNumber: string): Observable<{ sessionId: string; code: string }> {
    const code = this.generateCode();
    const formattedNumber = this.formatBangladeshNumber(phoneNumber);
    const message = `Your UnityMicroFund verification code is: ${code}. This code will expire in 10 minutes.`;

    return this.sendSms(formattedNumber, message).pipe(
      tap(() => {
        console.log('OTP sent to:', formattedNumber, 'Code:', code);
      }),
      catchError((error) => {
        console.error('OTP send error:', error);
        return throwError(() => error);
      }),
      () => new Observable<{ sessionId: string; code: string }>((subscriber) => {
        subscriber.next({
          sessionId: `session_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
          code: code,
        });
        subscriber.complete();
      })
    );
  }

  private formatBangladeshNumber(phone: string): string {
    let number = phone.replace(/[\s\-\(\)]/g, '');
    
    if (number.startsWith('+880')) {
      number = number.substring(4);
    } else if (number.startsWith('880')) {
      number = number.substring(3);
    } else if (number.startsWith('0')) {
      number = number.substring(1);
    }
    
    if (!number.startsWith('880')) {
      number = '880' + number;
    }
    
    return number;
  }

  private generateCode(): string {
    return Math.floor(100000 + Math.random() * 900000).toString();
  }

  isConfigured(): boolean {
    return !!this.config.apiKey;
  }
}

File: UnityMicroFund/UnityMicroFund.API/Infrastructure/Sms/SmsService.cs
using System.Net.Http.Json;

namespace UnityMicroFund.API.Infrastructure.Sms;

public interface ISmsService
{
    bool IsConfigured { get; }
    Task<bool> SendSmsAsync(string phone, string message, CancellationToken cancellationToken = default);
}

public class SmsService : ISmsService
{
    private readonly HttpClient _httpClient;
    private readonly IConfiguration _configuration;
    private readonly ILogger<SmsService> _logger;

    public SmsService(HttpClient httpClient, IConfiguration configuration, ILogger<SmsService> logger)
    {
        _httpClient = httpClient;
        _configuration = configuration;
        _logger = logger;
    }

    public bool IsConfigured => !string.IsNullOrEmpty(_configuration["Sms:ApiKey"]);

    public async Task<bool> SendSmsAsync(string phone, string message, CancellationToken cancellationToken = default)
    {
        var sms = _configuration.GetSection("Sms");
        var apiKey = sms["ApiKey"];
        if (string.IsNullOrEmpty(apiKey))
        {
            _logger.LogWarning("SMS not configured. Skipping send to {Phone}", phone);
            return false;
        }

        var apiUrl = sms["ApiUrl"] ?? "https://app.seasms.com/api/v3";
        var senderId = sms["SenderId"] ?? "UnityMF";

        try
        {
            using var request = new HttpRequestMessage(HttpMethod.Post, $"{apiUrl.TrimEnd('/')}/sms/send")
            {
                Content = JsonContent.Create(new
                {
                    recipient = FormatBangladeshNumber(phone),
                    sender_id = senderId,
                    type = "plain",
                    message
                })
            };
            request.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", apiKey);

            var response = await _httpClient.SendAsync(request, cancellationToken);
            if (!response.IsSuccessStatusCode)
            {
                _logger.LogError("SMS send failed with status {Status} for {Phone}", response.StatusCode, phone);
                return false;
            }

            _logger.LogInformation("SMS sent successfully to {Phone}", phone);
            return true;
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to send SMS to {Phone}", phone);
            return false;
        }
    }

    private static string FormatBangladeshNumber(string phone)
    {
        var number = new string(phone.Where(char.IsDigit).ToArray());

        if (number.StartsWith("880"))
        {
            return number;
        }
        if (number.StartsWith("0"))
        {
            number = number[1..];
        }
        return "880" + number;
    }
}
